h2quic/client.go

```diff
@@ -44,57 +46,34 @@ var _ h2quicClient = &Client{}
 
 // NewClient creates a new client
 func NewClient(t *QuicRoundTripper, tlsConfig *tls.Config, hostname string) *Client {
-	c := &Client{
+	return &Client{
 		t:               t,
+		dialAddr:        quic.DialAddr,
 		hostname:        authorityAddr("https", hostname),
 		responses:       make(map[protocol.StreamID]chan *http.Response),
 		encryptionLevel: protocol.EncryptionUnencrypted,
+		config: &quic.Config{
+			TLSConfig:                     tlsConfig,
+			RequestConnectionIDTruncation: true,
+		},
+		handshakeChan: make(chan struct{}),
 	}
-	c.cryptoChangedCond = sync.Cond{L: &c.mutex}
-	c.config = &quic.Config{
-		ConnState:                     c.connStateCallback,
-		TLSConfig:                     tlsConfig,
-		RequestConnectionIDTruncation: true,
-	}
-	return c
 }
 
 // Dial dials the connection
-func (c *Client) Dial() error {
-	_, err := quic.DialAddr(c.hostname, c.config)
-	return err
-}
+func (c *Client) Dial() (err error) {
+	defer func() {
+		c.mutex.Lock()
```

Comment:
I don't think you actually need the mutex here if you always set the error before closing the chan.